app/src/main/java/org/dhis2/utils/customviews/YesNoView.java

```diff
@@ -51,13 +67,95 @@ public void init(Context context) {
         super.init(context);
     }
 
-    public void setValueType(ValueType valueType) {
+    public void setValueListener(OnValueChanged listener) {
+        this.valueListener = listener;
+    }
 
-        if (valueType == ValueType.TRUE_ONLY)
+    public void setValueType(ValueType valueType) {
+        this.valueType = valueType;
+        if (valueType == ValueType.TRUE_ONLY) {
             no.setVisibility(View.GONE);
-        else
+            checkNo.setVisibility(View.GONE);
+        } else {
             no.setVisibility(View.VISIBLE);
+            checkNo.setVisibility(View.VISIBLE);
+        }
+    }
 
+    public void setRendering(ValueTypeRenderingType rendering) {
+        switch (rendering) {
+            case HORIZONTAL_CHECKBOXES:
+                checksLayout.setVisibility(View.VISIBLE);
+                radioGroup.setVisibility(View.GONE);
+                checkGroup.setVisibility(View.VISIBLE);
+                yesOnlyToggle.setVisibility(View.GONE);
+                clearButton.setVisibility(View.GONE);
+                checkGroup.setOrientation(LinearLayout.HORIZONTAL);
+                radioGroup.setOnCheckedChangeListener(null);
+                checkYes.setOnCheckedChangeListener(checkBoxClickListener);
+                checkNo.setOnCheckedChangeListener(checkBoxClickListener);
+                yesOnlyToggle.setOnCheckedChangeListener(null);
+                break;
+            case VERTICAL_CHECKBOXES:
+                checksLayout.setVisibility(View.VISIBLE);
+                radioGroup.setVisibility(View.GONE);
+                checkGroup.setVisibility(View.VISIBLE);
+                yesOnlyToggle.setVisibility(View.GONE);
+                clearButton.setVisibility(View.GONE);
+                checkGroup.setOrientation(LinearLayout.VERTICAL);
+                radioGroup.setOnCheckedChangeListener(null);
+                checkYes.setOnCheckedChangeListener(checkBoxClickListener);
+                checkNo.setOnCheckedChangeListener(checkBoxClickListener);
+                yesOnlyToggle.setOnCheckedChangeListener(null);
+                break;
+            case VERTICAL_RADIOBUTTONS:
+                checksLayout.setVisibility(View.VISIBLE);
+                radioGroup.setVisibility(View.VISIBLE);
+                checkGroup.setVisibility(View.GONE);
+                yesOnlyToggle.setVisibility(View.GONE);
+                clearButton.setVisibility(View.VISIBLE);
+                radioGroup.setOrientation(LinearLayout.VERTICAL);
+                radioGroup.setOnCheckedChangeListener(radioClickListener);
+                checkYes.setOnCheckedChangeListener(null);
+                checkNo.setOnCheckedChangeListener(null);
```

Comment:
is it really needed to add null? or we can avoid all these initialization?

Reply:
Yes, to avoid loops when updating the value